Fixes Issue #248

jwishnie opened the pull request:
close 248

Updates all list views to new toggle logic, which is centralized in a new utility class ListViewUtils

jwishnie committed:
Addresses issue #248
https://github.com/opendatakit/collect/issues/248

Adds new utility class 'ListViewUtils' with static helpers for checking and unchecking ListView items.

NOTE: Behavior of 'toggle' button is not actually a toggle. It does not flip the state of individual items.
Instead what is implemented is really two binary functions "Check All" and "Uncheck All".

A clean reimplementation would replace the label of the 'Toggle All' button with alternating 'Check All' and 'Uncheck All'
labels, dependent on the current state of checked items in the list.

This will be filed as a new bug.

Changes to be committed:
	modified:   collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java
	new file:   collect_app/src/main/java/org/odk/collect/android/utilities/ListViewUtils.java

jwishnie committed:
Updated all ListViews that have Toggle buttons to use logic in ListViewUtils--previously they all had
their own logic from what looks like some cut-and-pasted code, with some differences.

NOTE: Can be refactored further especially the logic that resyncs the state of the model to the
newly all-checked or all-unchecked ListView

Changes to be committed:
	modified:   collect_app/src/main/java/org/odk/collect/android/activities/DataManagerList.java
	modified:   collect_app/src/main/java/org/odk/collect/android/activities/FormManagerList.java
	modified:   collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderList.java
	modified:   collect_app/src/main/java/org/odk/collect/android/utilities/ListViewUtils.java

jwishnie committed:
ignore emacs backup files

jwishnie committed:
Removed unneeded 'mToggled' state from list views

Changes to be committed:
	modified:   collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java
	modified:   collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderList.java

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java:
```diff
@@ -116,7 +117,7 @@
     private ArrayList<HashMap<String, String>> mFormList;
 
-    private boolean mToggled = false;
+    // private boolean mToggled = false;
```
We get history through git so I think code that's no longer needed should be deleted.

jwishnie replied:
Cool. I pulled the commented lines from a later commit that I can create a PR from. I'll wait to hear your response on the comments below and then clean everything up at once.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java:
```diff
@@ -192,7 +182,7 @@ public void onClick(View v) {
 
             // indicating whether or not select-all is on or off.
```
I believe this made sense when "toggle" meant a true alternation but isn't relevant anymore. The whole conditional and comment should be removed.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java:
```diff
@@ -286,7 +276,7 @@ private void clearChoices() {
     protected void onListItemClick(ListView l, View v, int position, long id) {
         super.onListItemClick(l, v, position, id);
-        mDownloadButton.setEnabled(!(selectedItemCount() == 0));
+        mDownloadButton.setEnabled(!(ListViewUtils.selectedItemCount(getListView()) == 0));
```
I think using != would make this easier to think about. Actually, I think > 0 makes even more sense!

jwishnie replied:
I intentionally made as few changes to the previous code as possible. In this case, specifically just substituted the refactored (moved-to-ListViewUtils) method and kept the check the same as before.

Your suggestion is a good one, but probably should be carried out throughout the code base in a different commit, no?

jwishnie replied:
But I agree! !=0 or > 0 is clearer!

lognaturel replied:
Good point that especially in the absence of tests it makes sense to limit the number of changes made. Still, I'd prefer to see some of these smaller things that are easy to verify manually cleaned up while we have the code and its intent firmly in our minds! That will make the behemoth eventual refactor just a little bit easier/more pleasant. But if you don't mind, I'd like to get some more opinions in since this kind of thing will be coming up a lot. @yanokwa, @batkinson?

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java:
```diff
@@ -713,7 +687,7 @@ public void formListDownloadingComplete(HashMap<String, FormDetails> result) {
             selectSupersededForms();
             mFormListAdapter.notifyDataSetChanged();
-            mDownloadButton.setEnabled(!(selectedItemCount() == 0));
+            mDownloadButton.setEnabled(!(ListViewUtils.selectedItemCount(getListView()) == 0));
```
Same as earlier, makes more sense to use > 0 or even to create something like a hasSelectedItems(lv) predicate.

jwishnie replied:
Same response as earlier. By intent. Change prior code base as little as possible unless undertaking a comprehensive cleanup (which wasn't my goal!)

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java:
```diff
@@ -341,8 +331,9 @@ private void downloadFormList() {
         super.onSaveInstanceState(outState);
-        outState.putBoolean(BUNDLE_TOGGLED_KEY, mToggled);
-        outState.putInt(BUNDLE_SELECTED_COUNT, selectedItemCount());
+        // TODO: Is it safe to remove BUNDLE_TOGGLED_KEY ?
```
It shouldn't be a problem to remove all references to it. Do you mean that you want to actually remove the key? I'd tend to just leave it.

jwishnie replied:
I mean is it safe to remove the writing out and reading back in of the value. 

Though I don't see a need to keep the constant around either if it is not used.

Your opinion?

lognaturel replied:
I don't see any problem with removing the writing out and reading back in but maybe I'm missing something. What do you think the issue could be? Removing the BUNDLE_TOGGLED_KEY constant is also fine as it's just an identifier for the key.

I was talking about removing the persisted key/value pair but actually I don't think `onSaveInstanceState` persists values beyond app termination.

jwishnie replied:
`mToggled` is no longer used, so I don't see any reason to bother saving the state.

My concern was that I don't know how Java/Android serialization works. In the bad-old-days of  ObjectiveC, when objects were serialized, they wrote out an opaque binary blob.

If you changed the _interpretation_ of that by changing what values, or the order of those values, are read in on de-serialization, it would break and even could lead to smashing the stack and code-injection...

From the use of _keys_, I'm guessing this serialization is smarter and is writing out key/value pairs, but I didn't want to muck without asking because in dumb object-graph serialization, that's a big no-no.

lognaturel replied:
Ooh, that's also exciting! Yeah, none of that here. 😊

lognaturel reviewed (changes requested):
Welcome @jwishnie! 🎉 This is much more sane behavior and code structure!

I made a few small notes inline.

lognaturel commented:
It's too bad that `mToggled`, `mSelectedCount` and `mSelected` have to exist at all. [getCheckedItemIds](https://developer.android.com/reference/android/widget/AbsListView.html#getCheckedItemIds()) could replace them today but it would probably be ideal to be able to use [getCheckedItemCount()](https://developer.android.com/reference/android/widget/AbsListView.html#getCheckedItemCount()), only available in API 11+ (see #293). I'm curious -- for your `ListViewUtils.selectedItemCount()` why did you iterate through the position sparse array rather than getting the length of the ID array? The latter seems a little easier to read to me.

jwishnie commented:
re: ListViewUtils.selectedItemCount() -- this code was copy as-was from the FormDownloadManager. I'm not super familiar with the Android APIs, but my understanding of how the getSelectedItemCount and the Sparse Array works is that:

1.  size() when called on a sparse array is the number of items in the sparse array, which can be anywhere between the number of 'true' values and the total number of possible values, but is definitely not guaranteed to be the total number of possible values.

2. it was unclear to me that it is even guaranteed to be always the number of _true_ values and may in fact be the number of slots that have been allocated (whether true, or set later to false) during normal use.

lognaturel commented:
I agree with your interpretation of [`getCheckedIntemPositions()`](https://developer.android.com/reference/android/widget/AbsListView.html#getCheckedItemPositions()). I'm just wondering why you didn't use [`getCheckedItemIds()`](https://developer.android.com/reference/android/widget/AbsListView.html#getCheckedItemIds()) which returns a `long[]`!

lognaturel commented:
Oh, I see, you're saying it's in the same vein as changing !(A == 0) to A > 0 and maybe doesn't belong in this PR because it's a change to the code structure. Let's see what other opinions we get on that and we can decide where to make this change.

jwishnie committed:
Converted occurances of !(count == 0) to count > 0, e.g.:
- mUploadButton.setEnabled(!(mSelected.size() == 0));
+ mUploadButton.setEnabled(mSelected.size() > 0);

Changes to be committed:
	modified:   collect_app/src/main/java/org/odk/collect/android/activities/DataManagerList.java
	modified:   collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java
	modified:   collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderList.java

jwishnie commented:
I was looking into`getCheckedItemIds()` but that became a thread that I decided not to pull, as follows:

* Found places where ArrayLists<Long> where being used (can't remember where...) and thought they should just be long[] to take the return from getCheckedItemIds() directly without need to convert to Long[]s

* Started wondering why mSelceted (and it's brethren) existed at all and if they could be replaced with real time checks...

* stopped pulling on thread ;-)

jwishnie commented:
Question: I have a couple new commits that--
* remove the comments
* change checks from `!(count) == 0` to `count > 0`

Should I cancel this pull request and issue a new one with all the commits?

jwishnie commented:
**UPDATE**

Scratch what I said about `getCheckedItemIds()`! 

I didn't use it because it didn't work--just returned `null`

Again, a thread I didn't want to pull, but I suspect it has to do with this (from the Android documentation--emphasis is mine):

> Returns the set of checked items ids. The result is only valid if the choice mode has not been set to CHOICE_MODE_NONE **and the adapter has stable IDs. (hasStableIds() == true)**

I think the ListViews in question do _not_ have stable IDs, whatever that means.

Seemed outside the scope of the Issue at hand to go learn what that means...

lognaturel commented:
Ooooh, exciting. Ok, I think you're right to drop all this last stuff we've been talking about and to file an issue to remove `mSelected` etc and use Android APIs as possible. That can include investigating stable IDs.

I generally think it makes sense to keep changes as commits to the same branch/PR. What do you think?

jwishnie committed:
Removed unused BUNDLE_TOGGLED_KEY and calls to write it out, read it in.

Changes to be committed:
	modified:   collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java
	modified:   collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderList.java
	modified:   collect_app/src/main/java/org/odk/collect/android/utilities/ListViewUtils.java

jwishnie committed:
Added gradle.properties to .gitigore so that people can adjust local gradle settings without stomping on others.

Changes to be committed:
	modified:   .gitignore

jwishnie committed:
Gradle upgraded to 2.2.3

Changes to be committed:
	modified:   build.gradle
